File: src/aimformat/canonical.py
"""Canonical serialization and hashing (spec §11).

Byte-determinism is load-bearing: equality of serializations — not any
parser's opinion — is what verifies the history chain, and `doc_hash` anchors
checkpoints. The rules implemented here:

- attribute order: ``data-aim``/``data-aim-container``, ``id``, ``class``,
  ``style`` first; remaining attributes alphabetical; ``src``/``href`` last
- ``class`` tokens sorted alphabetically; inline style props in whitelist
  order, ``; ``-separated, no trailing semicolon
- text escapes ``& < >`` only; attribute values escape ``& "`` only
- foreign (SVG) attribute case re-adjusted (``viewBox`` …)
- one top-level construct per line — precisely: constructs never *share* a
  line (a construct with significant internal newlines, e.g. ``pre``, spans
  physical lines)
- typed scripts and the embedded stylesheet are block-laid-out (open tag /
  content lines / close tag); the theme block is a single hashed line
- JSON/JSONL: sorted keys, compact separators, raw UTF-8, ``</`` written
  ``<\\/``
- ``doc_hash`` = sha256 over the reduced projection: the ``<html …>`` open
  tag, the theme line, and each body content construct line, LF-joined with
  a trailing LF
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Iterable

from .dom import Comment, Element, Fragment, Nodeish, Text
from .registry import REGISTRY

LINE_CONTAINERS = frozenset({"html", "head", "body", "aim-proposals", "aim-assets"})


def escape_text(s: str) -> str:
    return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def escape_attr(s: str) -> str:
    return s.replace("&", "&amp;").replace('"', "&quot;")


def sort_class_tokens(value: str) -> str:
    """Class tokens sorted and de-duplicated (a set, canonically spelled)."""
    return " ".join(sorted(set(value.split())))


def normalize_style(value: str) -> str:
    """Inline style as a normal form: whitelist properties in registry
    order, later duplicates win, `; `-separated, no trailing semicolon.
    Unknown properties (a lint error anyway) keep authored order at the
    end so the violation stays visible rather than being reshuffled."""
    known: dict[str, str] = {}
    unknown: list[tuple[str, str]] = []
    for piece in value.split(";"):
        piece = piece.strip()
        if not piece or ":" not in piece:
            continue
        prop, val = (s.strip() for s in piece.split(":", 1))
        if prop in REGISTRY.style_prop_order:
            known[prop] = val
        else:
            unknown.append((prop, val))
    ordered = [(p, known[p]) for p in REGISTRY.style_prop_order if p in known]
    return "; ".join(f"{p}:{v}" for p, v in ordered + unknown)


def canonical_attrs(el: Element, *, in_svg: bool) -> str:
    adjust = REGISTRY.svg_case_adjust

    def fix(name: str) -> str:
        return adjust.get(name, name) if in_svg else name

    remaining: dict[str, str | None] = {}
    for k, v in el.attrs:
        # HTML semantics: the FIRST duplicate wins, matching Element.get —
        # last-wins here would rename a chunk under `aim normalize` and
        # break the events targeting the id the reader resolved
        remaining.setdefault(k, v)
    ordered: list[tuple[str, str | None]] = []
    for k in REGISTRY.attr_first:
        if k in remaining:
            ordered.append((k, remaining.pop(k)))
    tail = [(k, remaining.pop(k)) for k in REGISTRY.attr_last if k in remaining]
    ordered += sorted(remaining.items()) + tail

    parts = []
    for k, v in ordered:
        if k == "class" and v is not None:
            v = sort_class_tokens(v)
            if not v:
                continue  # empty class has no canonical spelling
        if k == "style" and v is not None:
            v = normalize_style(v)
            if not v:
                continue
        parts.append(fix(k) if v is None else f'{fix(k)}="{escape_attr(v)}"')
    return (" " + " ".join(parts)) if parts else ""


def _is_block_raw(el: Element) -> bool:
    if el.tag == "script" and "+" in (el.get("type") or ""):
        return True
    return el.tag == "style" and el.has("data-aim-css")


def _is_registry_svg(el: Element) -> bool:
    return el.tag == "svg" and el.has("aria-hidden")


def serialize(node: Nodeish, *, in_svg: bool = False) -> str:
    """Inline canonical serialization of one node (no trailing newline).

    A normal form, not an echo: HTML void elements never carry a slash
    however they were written, foreign (SVG-context) elements with no
    content always self-close, and every other element always has an
    explicit end tag (spec §11.1)."""
    if isinstance(node, Text):
        return escape_text(node.data)
    if isinstance(node, Comment):
        return f"<!--{node.data}-->"
    svg_here = in_svg or node.tag == "svg"
    open_tag = f"<{node.tag}{canonical_attrs(node, in_svg=svg_here)}"
    if node.tag in REGISTRY.void_elements and not svg_here:
        return open_tag + ">"
    if svg_here and not node.children and node.raw is None:
        return open_tag + "/>"
    if node.raw is not None:
        return f"{open_tag}>{node.raw}</{node.tag}>"
    inner = "".join(serialize(c, in_svg=svg_here) for c in node.children)
    return f"{open_tag}>{inner}</{node.tag}>"


def serialize_run(members: Iterable[Element]) -> str:
    """A chunk's serialization: its member elements concatenated in order."""
    return "".join(serialize(m) for m in members)


def _lines(node: Nodeish, *, in_svg: bool = False) -> list[str]:
    if isinstance(node, (Text, Comment)):
        s = serialize(node)
        return [s] if s.strip() else []
    svg_here = in_svg or node.tag == "svg"
    if node.tag in LINE_CONTAINERS or _is_registry_svg(node):
        out = [f"<{node.tag}{canonical_attrs(node, in_svg=svg_here)}>"]
        for c in node.children:
            out += _lines(c, in_svg=svg_here)
        out.append(f"</{node.tag}>")
        return out
    if node.raw is not None and _is_block_raw(node):
        body = node.raw
        body = body[1:] if body.startswith("\n") else body
        body = body[:-1] if body.endswith("\n") else body
        return (
            [f"<{node.tag}{canonical_attrs(node, in_svg=svg_here)}>"]
            + (body.split("\n") if body else [])
            + [f"</{node.tag}>"]
        )
    return [serialize(node, in_svg=in_svg)]


def document_text(fragment: Fragment) -> str:
    """Full canonical text of a parsed document."""
    lines = [f"<!{fragment.doctype}>"] if fragment.doctype else []
    for child in fragment.children:
        lines += _lines(child)
    return "\n".join(lines) + "\n"


# --------------------------------------------------------------------------
# JSON canonical form
def canonical_json(obj: object) -> str:
    """RFC 8785-flavoured canonical JSON, safe to embed in a script block."""
    return json.dumps(obj, ensure_ascii=False, sort_keys=True, separators=(",", ":")).replace(
        "</", "<\\/"
    )


# --------------------------------------------------------------------------
# hashing
def sha256_prefixed(data: str | bytes) -> str:
    if isinstance(data, str):
        data = data.encode("utf-8")
    return "sha256:" + hashlib.sha256(data).hexdigest()


def doc_hash(
    html_open_line: str,
    theme_line: str | None,
    construct_lines: Iterable[str],
    *,
    doc_settings_line: str | None = None,
) -> str:
    """The reduced-projection hash anchoring checkpoints (spec §11.4).

    The settings block participates only when present, so documents without
    one hash exactly as they did before it existed."""
    lines = [html_open_line]
    if doc_settings_line:
        lines.append(doc_settings_line)
    if theme_line:
        lines.append(theme_line)
    lines.extend(construct_lines)
    return sha256_prefixed("\n".join(lines) + "\n")

File: ts/src/canonical.ts
/**
 * Canonical serialization and hashing (spec §11) — a direct port of the
 * Python SDK's `canonical.py`. Byte-determinism is load-bearing: `Chunk.html`
 * and `docHash` are defined as *these* bytes, and the parity goldens pin the
 * two implementations to each other field by field.
 */

import { Comment, Element, Text, type Nodeish } from "./dom.ts";
import {
  ATTR_FIRST,
  ATTR_LAST,
  STYLE_PROP_ORDER,
  SVG_CASE_ADJUST,
  VOID_ELEMENTS,
} from "./registry.data.ts";
import { sha256Prefixed } from "./sha256.ts";

export function escapeText(s: string): string {
  return s
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
}

export function escapeAttr(s: string): string {
  return s.replaceAll("&", "&amp;").replaceAll('"', "&quot;");
}

/**
 * Compare strings by Unicode code point, matching Python's `str` ordering.
 * JS `<` compares UTF-16 code units, which ranks astral characters (encoded
 * as surrogate pairs, 0xD800–0xDFFF) below BMP characters above U+D7FF —
 * the opposite of their code-point order.
 */
export function compareCodePoints(a: string, b: string): number {
  let i = 0;
  while (i < a.length && i < b.length) {
    const ca = a.codePointAt(i)!;
    const cb = b.codePointAt(i)!;
    if (ca !== cb) return ca - cb;
    i += ca > 0xffff ? 2 : 1;
  }
  return a.length - i === 0 ? (b.length - i === 0 ? 0 : -1) : 1;
}

/** Class tokens sorted and de-duplicated (a set, canonically spelled). */
export function sortClassTokens(value: string): string {
  return [...new Set(value.split(/\s+/).filter((t) => t.length > 0))]
    .sort(compareCodePoints)
    .join(" ");
}

/**
 * Inline style as a normal form: whitelist properties in registry order,
 * later duplicates win, `; `-separated, no trailing semicolon. Unknown
 * properties (a lint error anyway) keep authored order at the end so the
 * violation stays visible rather than being reshuffled.
 */
export function normalizeStyle(value: string): string {
  const known = new Map<string, string>();
  const unknown: Array<[string, string]> = [];
  for (const piece of value.split(";")) {
    const trimmed = piece.trim();
    if (trimmed.length === 0 || !trimmed.includes(":")) continue;
    const colon = trimmed.indexOf(":");
    const prop = trimmed.slice(0, colon).trim();
    const val = trimmed.slice(colon + 1).trim();
    if (STYLE_PROP_ORDER.includes(prop)) known.set(prop, val);
    else unknown.push([prop, val]);
  }
  const ordered: Array<[string, string]> = [];
  for (const prop of STYLE_PROP_ORDER) {
    const val = known.get(prop);
    if (val !== undefined) ordered.push([prop, val]);
  }
  return [...ordered, ...unknown].map(([p, v]) => `${p}:${v}`).join("; ");
}

export function canonicalAttrs(el: Element, inSvg: boolean): string {
  const fix = (name: string): string =>
    inSvg ? (SVG_CASE_ADJUST[name] ?? name) : name;

  const remaining = new Map<string, string | null>();
  for (const [k, v] of el.attrs) {
    // HTML semantics: the FIRST duplicate wins, matching Element.get —
    // last-wins here would rename a chunk under `aim normalize` and
    // break the events targeting the id the reader resolved
    if (!remaining.has(k)) remaining.set(k, v);
  }
  const ordered: Array<[string, string | null]> = [];
  for (const k of ATTR_FIRST) {
    if (remaining.has(k)) {
      ordered.push([k, remaining.get(k)!]);
      remaining.delete(k);
    }
  }
  const tail: Array<[string, string | null]> = [];
  for (const k of ATTR_LAST) {
    if (remaining.has(k)) {
      tail.push([k, remaining.get(k)!]);
      remaining.delete(k);
    }
  }
  ordered.push(
    ...[...remaining.entries()].sort(([a], [b]) => compareCodePoints(a, b)),
  );
  ordered.push(...tail);

  const parts: string[] = [];
  for (const [k, rawValue] of ordered) {
    let v = rawValue;
    if (k === "class" && v !== null) {
      v = sortClassTokens(v);
      if (v.length === 0) continue; // empty class has no canonical spelling
    }
    if (k === "style" && v !== null) {
      v = normalizeStyle(v);
      if (v.length === 0) continue;
    }
    parts.push(v === null ? fix(k) : `${fix(k)}="${escapeAttr(v)}"`);
  }
  return parts.length > 0 ? ` ${parts.join(" ")}` : "";
}

/**
 * Inline canonical serialization of one node (no trailing newline).
 *
 * A normal form, not an echo: HTML void elements never carry a slash
 * however they were written, and foreign (SVG-context) elements with no
 * content always self-close (spec §11.1).
 */
export function serialize(node: Nodeish, inSvg = false): string {
  if (node instanceof Text) return escapeText(node.data);
  if (node instanceof Comment) return `<!--${node.data}-->`;
  const svgHere = inSvg || node.tag === "svg";
  const openTag = `<${node.tag}${canonicalAttrs(node, svgHere)}`;
  if (VOID_ELEMENTS.has(node.tag) && !svgHere) return `${openTag}>`;
  if (svgHere && node.children.length === 0 && node.raw === null)
    return `${openTag}/>`;
  if (node.selfClosing) return `${openTag}/>`;
  if (node.raw !== null) return `${openTag}>${node.raw}</${node.tag}>`;
  const inner = node.children.map((c) => serialize(c, svgHere)).join("");
  return `${openTag}>${inner}</${node.tag}>`;
}

/** A chunk's serialization: its member elements concatenated in order. */
export function serializeRun(members: readonly Element[]): string {
  return members.map((m) => serialize(m)).join("");
}

/**
 * The reduced-projection hash anchoring checkpoints (spec §11.3): the
 * `<html …>` open line, the settings-block line (when present), the theme
 * line (when present), and each body content construct line, LF-joined
 * with a trailing LF.
 */
export function docHash(
  htmlOpenLine: string,
  themeLine: string | null,
  constructLines: readonly string[],
  docSettingsLine: string | null = null,
): string {
  const lines = [htmlOpenLine];
  if (docSettingsLine !== null && docSettingsLine.length > 0)
    lines.push(docSettingsLine);
  if (themeLine !== null && themeLine.length > 0) lines.push(themeLine);
  lines.push(...constructLines);
  return sha256Prefixed(`${lines.join("\n")}\n`);
}
